blocks/soapy/include/gnuradio-4.0/soapy/SoapyRaiiWrapper.hpp

```diff
@@ -0,0 +1,805 @@
+
+    /*******************************************************************
+     * Time API
+     ******************************************************************/
+
+    std::vector<std::string> listAvailableTimeSources() const {
+        std::size_t length = 0UZ;
+        return detail::convertToCpp(SoapySDRDevice_listTimeSources(_device.get(), &length), length);
+    }
+
+    std::string getTimeSource() const {
+        char*       value = SoapySDRDevice_getTimeSource(_device.get());
+        std::string result(value);
+        free(value);
+        return result;
+    }
+
+    void setTimeSource(const std::string& source, std::source_location location = std::source_location::current()) {
+        if (int error = SoapySDRDevice_setTimeSource(_device.get(), source.c_str()); error) {
+            throw gr::exception(fmt::format("Soapy error ({}): {}", error, SoapySDR_errToStr(error)), location);
+        }
+    }
+
+    std::uint64_t getHardwareTime(std::string_view what = {}) const { return static_cast<std::uint64_t>(SoapySDRDevice_getHardwareTime(_device.get(), what.data())); }
+
+    void setHardwareTime(long long timeNs, const std::string& event = "", std::source_location location = std::source_location::current()) {
+        if (int error = SoapySDRDevice_setHardwareTime(_device.get(), timeNs, event.empty() ? nullptr : event.c_str()); error) {
+            throw gr::exception(fmt::format("Soapy error ({}): {}", error, SoapySDR_errToStr(error)), location);
+        }
+    }
+
+    /**
+     * Get the master clock rate of the device.
+     * \return the clock rate in Hz
+     */
+    double getMasterClockRate() const { return SoapySDRDevice_getMasterClockRate(_device.get()); }
+
+    template<typename TValueType, int direction>
+    class Stream {
+        std::shared_ptr<SoapySDRDevice> _device{nullptr};
+        std::shared_ptr<SoapySDRStream> _stream{nullptr};
+
+    protected:
+        Stream(std::shared_ptr<SoapySDRDevice> device, SoapySDRStream* cStream)
+            : _device(device), _stream(cStream, [device](SoapySDRStream* stream) {
+                  if (device.get() == nullptr || stream == nullptr) {
+                      return;
+                  }
+                  if (int ret = SoapySDRDevice_closeStream(device.get(), stream); ret) {
+                      throw gr::exception(fmt::format("error({}) closing Device: '{}'", ret, SoapySDR_errToStr(ret)));
+                  }
+              }) {}
+        friend class Device;
+
+    public:
+        using value_type                       = TValueType;
+        constexpr static std::size_t Direction = direction;
+
+        Stream() noexcept                = default;
+        Stream(const Stream&)            = delete;
+        Stream& operator=(const Stream&) = delete;
+        Stream(Stream&& other) noexcept : Stream() { swap(*this, other); }
```

Comment:
While this is legal, as a user, it can have unexpected behavour:
- moved-from state is not empty
- moved from state contains old state of the thing we moved into